# ✈️ TravelTide – Mastery Project

# **Import Data**



To ensure smooth access for every notebook user and to avoid issues with large local files, I created a Lakehouse on AWS S3.  
This notebook loads the publicly available datasets directly from the **Bronze layer** of that Lakehouse.

**Goal:**

- Load flight, hotel, session, and user data for initial exploration

⏱️ *Note: Loading all datasets may take up to 3 minutes depending on your connection.*


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Plot style
sns.set(style="whitegrid")
plt.rcParams["figure.figsize"] = (12, 6)

def load_flights():
    url = "https://lakehouse-masteryproject-2025.s3.eu-north-1.amazonaws.com/bronze/public_flights_export_2025-03-31_134734.csv"
    return pd.read_csv(url)

def load_hotels():
    url = "https://lakehouse-masteryproject-2025.s3.eu-north-1.amazonaws.com/bronze/public_hotels_export_2025-03-31_171805.csv"
    return pd.read_csv(url)

def load_sessions():
    url = "https://lakehouse-masteryproject-2025.s3.eu-north-1.amazonaws.com/bronze/public_sessions_export_2025-03-31_221253.csv"
    return pd.read_csv(url)

def load_users():
    url = "https://lakehouse-masteryproject-2025.s3.eu-north-1.amazonaws.com/bronze/public_users_export_2025-04-01_101058.csv"
    return pd.read_csv(url)

In [2]:
df_flights = load_flights()
df_hotels = load_hotels()
df_sessions = load_sessions()
df_users = load_users()

print("✅ Flight, hotel, session and user data loaded successfully.")

✅ Flight, hotel, session and user data loaded successfully.


# **EDA & Feature Engineering**

## 🔍 Data Overview | Cleansing | Feature Engineering


This section provides a structured overview of the dataset and prepares key features for further analysis.  
The focus is both on understanding the raw data and on deriving new, relevant features based on domain-specific rules:

- Dataset dimensions  
- Column names  
- Data types and missing values  
- First few sample rows  
- Statistical summary of all columns  
- Creation and validation of new features (Feature Engineering) using domain knowledge

### **Flights**

#### Flights - Overview

In [3]:
# 🧭 Quick overview of the dataset

print("🔹 SHAPE:")
print(df_flights.shape)  # Rows and columns

print("\n🔹 COLUMNS:")
print(df_flights.columns.tolist())  # List of column names

print("\n🔹 TYPES & NULLS:")
df_flights.info()  # Column types, non-null counts

print("\n🔹 SAMPLE DATA:")
display(df_flights.head())  # First few rows

print("\n🔹 BASIC STATS:")
display(df_flights.describe(include='all'))  # Summary for all types

🔹 SHAPE:
(1847300, 13)

🔹 COLUMNS:
['trip_id', 'origin_airport', 'destination', 'destination_airport', 'seats', 'return_flight_booked', 'departure_time', 'return_time', 'checked_bags', 'trip_airline', 'destination_airport_lat', 'destination_airport_lon', 'base_fare_usd']

🔹 TYPES & NULLS:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1847300 entries, 0 to 1847299
Data columns (total 13 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   trip_id                  object 
 1   origin_airport           object 
 2   destination              object 
 3   destination_airport      object 
 4   seats                    int64  
 5   return_flight_booked     bool   
 6   departure_time           object 
 7   return_time              object 
 8   checked_bags             int64  
 9   trip_airline             object 
 10  destination_airport_lat  float64
 11  destination_airport_lon  float64
 12  base_fare_usd            float64
dtypes: bool(1), float64(3)

,trip_id,origin_airport,destination,destination_airport,seats,return_flight_booked,departure_time,return_time,checked_bags,trip_airline,destination_airport_lat,destination_airport_lon,base_fare_usd
0,693092-7b8767746abd45219ed468d58da9f702,SJC,houston,EFD,1,True,2023-04-02 07:00:00,2023-04-06 07:00:00,0,United Airlines,29.607,-95.159,479.31
1,693124-1032149a9e01427da914b17ce6ef1926,DEN,montreal,YHU,1,True,2023-04-04 07:00:00,2023-04-17 07:00:00,0,Delta Air Lines,45.517,-73.417,575.78
2,693149-0fdb31c7dc114a4d8a318f0feada365c,YEG,seattle,SEA,2,True,2023-04-01 10:00:00,2023-04-07 10:00:00,0,Alaska Airlines,47.530,-122.302,336.72
3,693169-d120c3721d0b4bcca278639c981e4322,MCF,san antonio,SKF,1,True,2023-04-04 11:00:00,2023-04-05 11:00:00,0,United Airlines,29.529,-98.279,295.37
4,693181-f4a2363db0ee452f8ac8c0fe3d4ef0ad,YVR,new york,LGA,1,True,2023-03-31 07:00:00,2023-04-02 07:00:00,1,JetBlue Airways,40.640,-73.779,694.26



🔹 BASIC STATS:


,trip_id,origin_airport,destination,destination_airport,seats,return_flight_booked,departure_time,return_time,checked_bags,trip_airline,destination_airport_lat,destination_airport_lon,base_fare_usd
count,1847300,1847300,1847300,1847300,1.847300e+06,1847300,1847300,1761160,1.847300e+06,1847300,1.847300e+06,1.847300e+06,1.847300e+06
unique,1847300,159,140,219,NaN,2,17710,17964,NaN,355,NaN,NaN,NaN
top,577592-b92f730e10a74530b47f997a99de1497,JFK,new york,LGA,NaN,True,2023-03-18 07:00:00,2023-03-10 07:00:00,NaN,Delta Air Lines,NaN,NaN,NaN
freq,1,116524,249601,124948,NaN,1761160,2040,1902,NaN,265498,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,1.250766e+00,NaN,NaN,NaN,6.080139e-01,NaN,3.812317e+01,-8.328964e+01,6.442810e+02
std,NaN,NaN,NaN,NaN,6.401529e-01,NaN,NaN,NaN,7.170107e-01,NaN,8.056612e+00,4.447461e+01,1.096380e+03
min,NaN,NaN,NaN,NaN,0.000000e+00,NaN,NaN,NaN,0.000000e+00,NaN,-3.700800e+01,-1.579270e+02,0.000000e+00
25%,NaN,NaN,NaN,NaN,1.000000e+00,NaN,NaN,NaN,0.000000e+00,NaN,3.353500e+01,-1.123830e+02,2.027000e+02
50%,NaN,NaN,NaN,NaN,1.000000e+00,NaN,NaN,NaN,1.000000e+00,NaN,3.985800e+01,-8.775200e+01,3.960700e+02
75%,NaN,NaN,NaN,NaN,1.000000e+00,NaN,NaN,NaN,1.000000e+00,NaN,4.240900e+01,-7.377900e+01,6.374900e+02


#### Flights - Cleansing 

Ensure datetime format for time columns

In [4]:
# 1. Ensure datetime format for time columns
df_flights["departure_time"] = pd.to_datetime(df_flights["departure_time"], format="mixed", errors="coerce")
df_flights["return_time"] = pd.to_datetime(df_flights["return_time"], format="mixed", errors="coerce")

Standardize "Air Lines" to "Airlines"

In [25]:
df_flights['trip_airline'] = df_flights['trip_airline'].str.replace(r'\bAir\s+Lines\b', 'Airlines', regex=True)

#### Flights - Feature Engineering

Extracting the user_id out of the trip_id

In [26]:
df_flights['user_id'] = df_flights['trip_id'].str.extract(r'^([^-]+)')

In [27]:
df_flights = df_flights[
    ['trip_id', 'user_id', 'origin_airport', 'destination', 'destination_airport',
     'seats', 'return_flight_booked', 'departure_time', 'return_time',
     'checked_bags', 'trip_airline', 'destination_airport_lat',
     'destination_airport_lon', 'base_fare_usd']
]

df_flights


,trip_id,user_id,origin_airport,destination,destination_airport,seats,return_flight_booked,departure_time,return_time,checked_bags,trip_airline,destination_airport_lat,destination_airport_lon,base_fare_usd
0,693092-7b8767746abd45219ed468d58da9f702,693092,SJC,houston,EFD,1,True,2023-04-02 07:00:00,2023-04-06 07:00:00,0,United Airlines,29.607,-95.159,479.31
1,693124-1032149a9e01427da914b17ce6ef1926,693124,DEN,montreal,YHU,1,True,2023-04-04 07:00:00,2023-04-17 07:00:00,0,Delta Airlines,45.517,-73.417,575.78
2,693149-0fdb31c7dc114a4d8a318f0feada365c,693149,YEG,seattle,SEA,2,True,2023-04-01 10:00:00,2023-04-07 10:00:00,0,Alaska Airlines,47.530,-122.302,336.72
3,693169-d120c3721d0b4bcca278639c981e4322,693169,MCF,san antonio,SKF,1,True,2023-04-04 11:00:00,2023-04-05 11:00:00,0,United Airlines,29.529,-98.279,295.37
4,693181-f4a2363db0ee452f8ac8c0fe3d4ef0ad,693181,YVR,new york,LGA,1,True,2023-03-31 07:00:00,2023-04-02 07:00:00,1,JetBlue Airways,40.640,-73.779,694.26
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1847295,577017-b2a8f426280c4600b0785ab96288e1b3,577017,PHL,new york,LGA,1,True,2023-03-24 08:00:00,2023-03-25 08:00:00,0,KLM Royal Dutch Airlines,40.640,-73.779,28.01
1847296,577208-77ea1560346c44a380ca8d7294d4b990,577208,TLH,houston,IAH,1,True,2023-03-28 08:00:00,2023-03-31 08:00:00,0,American Airlines,29.607,-95.159,170.97
1847297,577450-99bf2f6662f64063ac316ed5ce8f3778,577450,DEN,houston,IAH,2,True,2023-03-25 07:00:00,2023-03-25 07:00:00,1,Singapore Airlines,29.607,-95.159,517.29
1847298,577510-350a2093a5b7429b8f7f1788f28c1508,577510,LGA,denver,DEN,1,True,2023-03-22 07:00:00,2023-03-26 07:00:00,0,United Airlines,39.858,-104.667,490.58


### **Hotels**

#### Hotels Overview

In [5]:
# 🧭 Quick overview of the dataset

print("🔹 SHAPE:")
print(df_hotels.shape)  # Rows and columns

print("\n🔹 COLUMNS:")
print(df_hotels.columns.tolist())  # List of column names

print("\n🔹 TYPES & NULLS:")
df_hotels.info()  # Column types, non-null counts

print("\n🔹 SAMPLE DATA:")
display(df_hotels.head())  # First few rows

print("\n🔹 BASIC STATS:")
display(df_hotels.describe(include='all'))  # Summary for all types

🔹 SHAPE:
(1918617, 7)

🔹 COLUMNS:
['trip_id', 'hotel_name', 'nights', 'rooms', 'check_in_time', 'check_out_time', 'hotel_per_room_usd']

🔹 TYPES & NULLS:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1918617 entries, 0 to 1918616
Data columns (total 7 columns):
 #   Column              Dtype 
---  ------              ----- 
 0   trip_id             object
 1   hotel_name          object
 2   nights              int64 
 3   rooms               int64 
 4   check_in_time       object
 5   check_out_time      object
 6   hotel_per_room_usd  int64 
dtypes: int64(3), object(4)
memory usage: 102.5+ MB

🔹 SAMPLE DATA:


,trip_id,hotel_name,nights,rooms,check_in_time,check_out_time,hotel_per_room_usd
0,234666-99574aa53f30402ba04bfe2f3337c566,NH Hotel - new york,2,1,2023-01-20 11:46:41.16,2023-01-22 11:00:00,226
1,234707-4c0cfdb5ff4a4cd8bda5feed22f790d7,Extended Stay - toronto,1,1,2023-01-19 15:38:40.11,2023-01-21 11:00:00,114
2,234745-a8f944dc121c4003b16c108c9c590965,Shangri-La - edmonton,3,1,2023-01-19 13:05:43.98,2023-01-23 11:00:00,322
3,234905-c1e14087ef1f4a7a8413a12ef4208135,NH Hotel - memphis,2,1,2023-01-16 11:50:56.985,2023-01-18 11:00:00,228
4,235101-cd0e752033db4f488fa75b5eb8feb2c8,Aman Resorts - jacksonville,1,1,2023-01-16 17:09:21.375,2023-01-18 11:00:00,163



🔹 BASIC STATS:


,trip_id,hotel_name,nights,rooms,check_in_time,check_out_time,hotel_per_room_usd
count,1918617,1918617,1.918617e+06,1.918617e+06,1918617,1918617,1.918617e+06
unique,1918617,2798,NaN,NaN,1317413,1221,NaN
top,234549-d7bfb13b8de1489783e00723207629a1,Extended Stay - new york,NaN,NaN,2023-03-19 11:00:00,2023-04-05 11:00:00,NaN
freq,1,14075,NaN,NaN,1190,6410,NaN
mean,NaN,NaN,4.271452e+00,1.189341e+00,NaN,NaN,1.773054e+02
std,NaN,NaN,4.488629e+00,4.879645e-01,NaN,NaN,1.171834e+02
min,NaN,NaN,-2.000000e+00,0.000000e+00,NaN,NaN,8.000000e+00
25%,NaN,NaN,1.000000e+00,1.000000e+00,NaN,NaN,9.900000e+01
50%,NaN,NaN,3.000000e+00,1.000000e+00,NaN,NaN,1.480000e+02
75%,NaN,NaN,6.000000e+00,1.000000e+00,NaN,NaN,2.220000e+02


#### Hotels Cleansing

**Calculated nights ≠ Actual nights spent in the room**

It's the number of billable nights based on hotel policy, not necessarily physical presence overnight.

In [6]:
from datetime import datetime, timedelta
import pandas as pd

# Hotel rules
CHECKIN_TIME = datetime.strptime("15:00", "%H:%M").time()
CHECKOUT_TIME = datetime.strptime("11:00", "%H:%M").time()
GRACE_PERIOD = timedelta(hours=1)  # Grace period: 1 hour after official checkout time

# Realistic nights calculation function based on hotel logic
def calculate_nights_realistic(row):
    check_in = row["check_in_time"]
    check_out = row["check_out_time"]

    if pd.isna(check_in) or pd.isna(check_out):
        return None  # Missing values

    check_in_date = check_in.date()
    check_out_date = check_out.date()

    # Base number of nights = difference in calendar days
    nights = (check_out_date - check_in_date).days

    # Checkout deadline including grace period
    checkout_deadline = datetime.combine(check_out_date, CHECKOUT_TIME) + GRACE_PERIOD

    if check_out > checkout_deadline:
        nights += 1  # Late checkout → extra night

    return nights

# 1. Ensure datetime format for time columns
df_hotels["check_in_time"] = pd.to_datetime(df_hotels["check_in_time"], format="mixed", errors="coerce")
df_hotels["check_out_time"] = pd.to_datetime(df_hotels["check_out_time"], format="mixed", errors="coerce")

# 2. Calculate nights using realistic hotel logic
df_hotels["nights_calculated"] = df_hotels.apply(calculate_nights_realistic, axis=1)

# 3. Compare with original 'nights' column
df_hotels["nights_diff"] = df_hotels["nights_calculated"] - df_hotels["nights"]

# 4. Show discrepancies if any
inconsistent = df_hotels[df_hotels["nights_diff"] != 0]

print("✅ First rows with calculated and reported nights:")
print(df_hotels[["check_in_time", "check_out_time", "nights", "nights_calculated", "nights_diff"]].head())

if not inconsistent.empty:
    print("\n⚠️ Bookings with discrepancies:")
    print(inconsistent[["check_in_time", "check_out_time", "nights", "nights_calculated", "nights_diff"]])
else:
    print("\n🎉 All bookings are correctly calculated!")


✅ First rows with calculated and reported nights:
            check_in_time      check_out_time  nights  nights_calculated  \
0 2023-01-20 11:46:41.160 2023-01-22 11:00:00       2                  2   
1 2023-01-19 15:38:40.110 2023-01-21 11:00:00       1                  2   
2 2023-01-19 13:05:43.980 2023-01-23 11:00:00       3                  4   
3 2023-01-16 11:50:56.985 2023-01-18 11:00:00       2                  2   
4 2023-01-16 17:09:21.375 2023-01-18 11:00:00       1                  2   

   nights_diff  
0            0  
1            1  
2            1  
3            0  
4            1  

⚠️ Bookings with discrepancies:
                  check_in_time      check_out_time  nights  \
1       2023-01-19 15:38:40.110 2023-01-21 11:00:00       1   
2       2023-01-19 13:05:43.980 2023-01-23 11:00:00       3   
4       2023-01-16 17:09:21.375 2023-01-18 11:00:00       1   
8       2023-01-17 19:47:43.395 2023-01-22 11:00:00       4   
9       2023-01-16 15:24:06.660 2023-01-21 

**Extracting Hotel Brand and City from Hotel Name Column**

In [7]:
# Split 'hotel_name' into two new columns: brand and city
df_hotels[["hotel_brand", "hotel_city"]] = df_hotels["hotel_name"].str.split(" - ", expand=True)

# Capitalize the first letter of each word in the city name
df_hotels["hotel_city"] = df_hotels["hotel_city"].str.title()



**Transform: df_hotels["hotel_per_room_usd"] into float type**

In [8]:
df_hotels["hotel_per_room_usd"] = df_hotels["hotel_per_room_usd"].astype(float)

### **Sessions**

#### Sessions Overview

In [9]:
# 🧭 Quick overview of the dataset

print("🔹 SHAPE:")
print(df_sessions.shape)  # Rows and columns

print("\n🔹 COLUMNS:")
print(df_sessions.columns.tolist())  # List of column names

print("\n🔹 TYPES & NULLS:")
df_sessions.info()  # Column types, non-null counts

print("\n🔹 SAMPLE DATA:")
display(df_sessions.head())  # First few rows

print("\n🔹 BASIC STATS:")
display(df_sessions.describe(include='all'))  # Summary for all types

🔹 SHAPE:
(5408063, 13)

🔹 COLUMNS:
['session_id', 'user_id', 'trip_id', 'session_start', 'session_end', 'flight_discount', 'hotel_discount', 'flight_discount_amount', 'hotel_discount_amount', 'flight_booked', 'hotel_booked', 'page_clicks', 'cancellation']

🔹 TYPES & NULLS:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5408063 entries, 0 to 5408062
Data columns (total 13 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   session_id              object 
 1   user_id                 int64  
 2   trip_id                 object 
 3   session_start           object 
 4   session_end             object 
 5   flight_discount         bool   
 6   hotel_discount          bool   
 7   flight_discount_amount  float64
 8   hotel_discount_amount   float64
 9   flight_booked           bool   
 10  hotel_booked            bool   
 11  page_clicks             int64  
 12  cancellation            bool   
dtypes: bool(5), float64(2), int64(2), object(4)
memory us

,session_id,user_id,trip_id,session_start,session_end,flight_discount,hotel_discount,flight_discount_amount,hotel_discount_amount,flight_booked,hotel_booked,page_clicks,cancellation
0,841803-95c93659053948049c73106ffdcf4df6,841803,841803-31d33d39b08f477bb8d9dd3b0dcbf1d3,2023-05-17 19:40:00,2023-05-17 19:43:48,False,False,NaN,NaN,True,True,30,False
1,841804-bf1bfe452c4d4372b04d9612134006e7,841804,NaN,2023-05-17 22:40:00,2023-05-17 22:41:09,False,False,NaN,NaN,False,False,9,False
2,841805-70d695ee6540481bae7750e368c10443,841805,841805-506d17340ffa41659736fd258138a9b5,2023-05-17 21:10:00,2023-05-17 21:13:03,False,False,NaN,NaN,True,True,24,False
3,841806-a03ac981d1c847979765669ca871a0cb,841806,NaN,2023-05-17 21:49:00,2023-05-17 21:54:09,True,True,0.2,0.2,False,False,42,False
4,841807-4febb023cd414a6196fc026af1928bfb,841807,NaN,2023-05-17 20:28:00,2023-05-17 20:28:29,False,False,NaN,NaN,False,False,4,False



🔹 BASIC STATS:


,session_id,user_id,trip_id,session_start,session_end,flight_discount,hotel_discount,flight_discount_amount,hotel_discount_amount,flight_booked,hotel_booked,page_clicks,cancellation
count,5408063,5.408063e+06,2335845,5408063,5408063,5408063,5408063,885796.000000,691380.000000,5408063,5408063,5.408063e+06,5408063
unique,5408063,NaN,2245175,960389,4868259,2,2,NaN,NaN,2,2,NaN,2
top,841802-0161005c612646c696d80acf427b4d0a,NaN,192187-8643d744ca5e404399b02c4cfe83c15b,2023-03-27 20:46:00,2023-03-26 11:23:37,False,False,NaN,NaN,False,False,NaN,False
freq,1,NaN,2,50,8,4431597,4626013,NaN,NaN,3416355,3398776,NaN,5317393
mean,NaN,3.988590e+05,NaN,NaN,NaN,NaN,NaN,0.139765,0.110950,NaN,NaN,1.876603e+01,NaN
std,NaN,2.707599e+05,NaN,NaN,NaN,NaN,NaN,0.085191,0.062021,NaN,NaN,2.218070e+01,NaN
min,NaN,0.000000e+00,NaN,NaN,NaN,NaN,NaN,0.050000,0.050000,NaN,NaN,0.000000e+00,NaN
25%,NaN,1.643270e+05,NaN,NaN,NaN,NaN,NaN,0.100000,0.050000,NaN,NaN,7.000000e+00,NaN
50%,NaN,3.641720e+05,NaN,NaN,NaN,NaN,NaN,0.100000,0.100000,NaN,NaN,1.500000e+01,NaN
75%,NaN,6.031520e+05,NaN,NaN,NaN,NaN,NaN,0.200000,0.150000,NaN,NaN,2.300000e+01,NaN


#### Sessions Cleansing

In [10]:

# 1. Ensure datetime format for time columns
df_sessions["session_start"] = pd.to_datetime(df_sessions["session_start"], format="mixed", errors="coerce")
df_sessions["session_end"] = pd.to_datetime(df_sessions["session_end"], format="mixed", errors="coerce")

In [11]:
df_sessions

,session_id,user_id,trip_id,session_start,session_end,flight_discount,hotel_discount,flight_discount_amount,hotel_discount_amount,flight_booked,hotel_booked,page_clicks,cancellation
0,841803-95c93659053948049c73106ffdcf4df6,841803,841803-31d33d39b08f477bb8d9dd3b0dcbf1d3,2023-05-17 19:40:00,2023-05-17 19:43:48,False,False,NaN,NaN,True,True,30,False
1,841804-bf1bfe452c4d4372b04d9612134006e7,841804,NaN,2023-05-17 22:40:00,2023-05-17 22:41:09,False,False,NaN,NaN,False,False,9,False
2,841805-70d695ee6540481bae7750e368c10443,841805,841805-506d17340ffa41659736fd258138a9b5,2023-05-17 21:10:00,2023-05-17 21:13:03,False,False,NaN,NaN,True,True,24,False
3,841806-a03ac981d1c847979765669ca871a0cb,841806,NaN,2023-05-17 21:49:00,2023-05-17 21:54:09,True,True,0.2,0.20,False,False,42,False
4,841807-4febb023cd414a6196fc026af1928bfb,841807,NaN,2023-05-17 20:28:00,2023-05-17 20:28:29,False,False,NaN,NaN,False,False,4,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...
5408058,841798-022fd195dce84aeca41492a209f4d456,841798,841798-67b1366799aa4659852a38bcedfb4fdc,2023-05-17 19:20:00,2023-05-17 19:21:14,False,False,NaN,NaN,False,True,10,False
5408059,841799-1852098e3f9c4dc084cf808ae3972338,841799,NaN,2023-05-17 02:05:00,2023-05-17 02:06:47,False,False,NaN,NaN,False,False,15,False
5408060,841800-24d4e9b0034643ca94ad6b08ebd694df,841800,841800-8710f272ab484cbd8654264d56370fc0,2023-05-17 08:14:00,2023-05-17 08:16:19,False,True,NaN,0.05,True,False,19,False
5408061,841801-ab1e9225713e44639c2238f7b4241662,841801,841801-e6338477bfc04ae1bd4e3d62b407bf1e,2023-05-17 11:30:00,2023-05-17 11:32:30,False,False,NaN,NaN,True,True,20,False


### **Users**

#### Users Overview

In [12]:
# 🧭 Quick overview of the dataset

print("🔹 SHAPE:")
print(df_users.shape)  # Rows and columns

print("\n🔹 COLUMNS:")
print(df_users.columns.tolist())  # List of column names

print("\n🔹 TYPES & NULLS:")
df_users.info()  # Column types, non-null counts

print("\n🔹 SAMPLE DATA:")
display(df_users.head())  # First few rows

print("\n🔹 BASIC STATS:")
display(df_users.describe(include='all'))  # Summary for all types


🔹 SHAPE:
(1020926, 11)

🔹 COLUMNS:
['user_id', 'birthdate', 'gender', 'married', 'has_children', 'home_country', 'home_city', 'home_airport', 'home_airport_lat', 'home_airport_lon', 'sign_up_date']

🔹 TYPES & NULLS:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020926 entries, 0 to 1020925
Data columns (total 11 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   user_id           1020926 non-null  int64  
 1   birthdate         1020926 non-null  object 
 2   gender            1020926 non-null  object 
 3   married           1020926 non-null  bool   
 4   has_children      1020926 non-null  bool   
 5   home_country      1020926 non-null  object 
 6   home_city         1020926 non-null  object 
 7   home_airport      1020926 non-null  object 
 8   home_airport_lat  1020926 non-null  float64
 9   home_airport_lon  1020926 non-null  float64
 10  sign_up_date      1020926 non-null  object 
dtypes: bool(2), float64(2), int64

,user_id,birthdate,gender,married,has_children,home_country,home_city,home_airport,home_airport_lat,home_airport_lon,sign_up_date
0,0,1990-01-22,F,False,False,usa,minneapolis,MSP,44.880,-93.217,2021-04-01
1,1,2000-11-08,M,False,False,usa,colorado springs,COS,38.806,-104.700,2021-04-01
2,2,1992-09-21,M,False,False,usa,portland,PDX,45.589,-122.597,2021-04-01
3,3,1996-11-27,F,False,False,usa,houston,IAH,29.980,-95.340,2021-04-01
4,4,1978-01-05,M,True,True,usa,honolulu,HNL,21.316,-157.927,2021-04-01



🔹 BASIC STATS:


,user_id,birthdate,gender,married,has_children,home_country,home_city,home_airport,home_airport_lat,home_airport_lon,sign_up_date
count,1.020926e+06,1020926,1020926,1020926,1020926,1020926,1020926,1020926,1.020926e+06,1.020926e+06,1020926
unique,NaN,22581,3,2,2,2,105,159,NaN,NaN,831
top,NaN,2006-09-03,M,False,False,usa,new york,JFK,NaN,NaN,2023-05-15
freq,NaN,169,558986,626109,701289,848354,121290,60711,NaN,NaN,3326
mean,5.104625e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.855352e+01,-9.392036e+01,NaN
std,2.947161e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.120248e+00,1.809645e+01,NaN
min,0.000000e+00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.131600e+01,-1.579270e+02,NaN
25%,2.552312e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.394200e+01,-1.120080e+02,NaN
50%,5.104625e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.929700e+01,-8.997700e+01,NaN
75%,7.656938e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.227600e+01,-7.745600e+01,NaN


.describe

In [13]:
df_users.describe(include="all")

,user_id,birthdate,gender,married,has_children,home_country,home_city,home_airport,home_airport_lat,home_airport_lon,sign_up_date
count,1.020926e+06,1020926,1020926,1020926,1020926,1020926,1020926,1020926,1.020926e+06,1.020926e+06,1020926
unique,NaN,22581,3,2,2,2,105,159,NaN,NaN,831
top,NaN,2006-09-03,M,False,False,usa,new york,JFK,NaN,NaN,2023-05-15
freq,NaN,169,558986,626109,701289,848354,121290,60711,NaN,NaN,3326
mean,5.104625e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.855352e+01,-9.392036e+01,NaN
std,2.947161e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6.120248e+00,1.809645e+01,NaN
min,0.000000e+00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.131600e+01,-1.579270e+02,NaN
25%,2.552312e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.394200e+01,-1.120080e+02,NaN
50%,5.104625e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.929700e+01,-8.997700e+01,NaN
75%,7.656938e+05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.227600e+01,-7.745600e+01,NaN


In [14]:
df_users.shape

(1020926, 11)

#### Users Cleansing

**Convert Signup Column to Datetime Format**

In [15]:
# 1. Ensure datetime format for time columns
df_users["sign_up_date"] = pd.to_datetime(df_users["sign_up_date"], format="mixed", errors="coerce")


**Format Birthdate Column**

In [16]:
df_users["birthdate_str"] = df_users["birthdate"].dt.strftime("%Y-%m-%d")



AttributeError: Can only use .dt accessor with datetimelike values

In [ ]:
df_users

,user_id,birthdate,gender,married,has_children,home_country,home_city,home_airport,home_airport_lat,home_airport_lon,sign_up_date
0,0,1990-01-22,F,False,False,usa,minneapolis,MSP,44.880,-93.217,2021-04-01
1,1,2000-11-08,M,False,False,usa,colorado springs,COS,38.806,-104.700,2021-04-01
2,2,1992-09-21,M,False,False,usa,portland,PDX,45.589,-122.597,2021-04-01
3,3,1996-11-27,F,False,False,usa,houston,IAH,29.980,-95.340,2021-04-01
4,4,1978-01-05,M,True,True,usa,honolulu,HNL,21.316,-157.927,2021-04-01
...,...,...,...,...,...,...,...,...,...,...,...
1020921,1020921,2005-05-22,M,False,True,usa,shreveport,BAD,32.502,-93.663,2023-07-20
1020922,1020922,1985-06-15,M,True,False,usa,oklahoma city,TIK,35.415,-97.386,2023-07-20
1020923,1020923,1995-09-21,F,False,False,canada,toronto,YKZ,43.862,-79.370,2023-07-20
1020924,1020924,2006-11-19,F,False,False,usa,lincoln,LNK,40.851,-96.759,2023-07-20


## 🧪 Data Quality & Integrity

This section checks for common data quality issues such as:
- Missing values
- Duplicate rows

### -flights

In [ ]:
# 🔍 Missing values per column
print("🔹 Missing values:")
print(df_flights.isnull().sum())

# 🔍 Duplicate rows
print("\n🔹 Number of duplicate rows:")
print(df_flights.duplicated().sum())


🔹 Missing values:
trip_id                        0
origin_airport                 0
destination                    0
destination_airport            0
seats                          0
return_flight_booked           0
departure_time                 0
return_time                86140
checked_bags                   0
trip_airline                   0
destination_airport_lat        0
destination_airport_lon        0
base_fare_usd                  0
dtype: int64

🔹 Number of duplicate rows:
0


**🧾 Summary of Data Quality Checks**

- ✅ No missing values in most columns.
- ⚠️ `return_time` has **86,140 missing values**, which likely indicates one-way trips without return flights.
- ✅ No duplicate rows detected.
- ✅ All other fields appear complete.

→ The missing `return_time` values are not necessarily errors but may reflect actual business logic (one-way bookings).

### - hotels

In [ ]:
# 🔍 Missing values per column
print("🔹 Missing values:")
print(df_hotels.isnull().sum())

# 🔍 Duplicate rows
print("\n🔹 Number of duplicate rows:")
print(df_hotels.duplicated().sum())

🔹 Missing values:
trip_id               0
hotel_name            0
nights                0
rooms                 0
check_in_time         0
check_out_time        0
hotel_per_room_usd    0
dtype: int64

🔹 Number of duplicate rows:
0


### - sessions

In [ ]:
# 🔍 Missing values per column
print("🔹 Missing values:")
print(df_sessions.isnull().sum())

# 🔍 Duplicate rows
print("\n🔹 Number of duplicate rows:")
print(df_sessions.duplicated().sum())

🔹 Missing values:
session_id                      0
user_id                         0
trip_id                   3072218
session_start                   0
session_end                     0
flight_discount                 0
hotel_discount                  0
flight_discount_amount    4522267
hotel_discount_amount     4716683
flight_booked                   0
hotel_booked                    0
page_clicks                     0
cancellation                    0
dtype: int64

🔹 Number of duplicate rows:
0


**🧾 Data Quality Summary – Session Data**

- ✅ No duplicate rows.
- ✅ No missing values in session core fields such as `session_id`, `user_id`, `session_start`, `session_end`, and booking indicators.
- ⚠️ `trip_id` has over **3 million missing values**, which likely indicates that many sessions did not lead to a trip booking.
- ⚠️ `flight_discount_amount` and `hotel_discount_amount` have over **4.5 million missing values**, possibly reflecting sessions where no discount was applied.

→ These missing values are likely **structural** rather than invalid.  
They may represent **non-converted sessions** or **visits without discounts** and should be treated accordingly during feature engineering.

### - users

In [ ]:
# 🔍 Missing values per column
print("🔹 Missing values:")
print(df_users.isnull().sum())

# 🔍 Duplicate rows
print("\n🔹 Number of duplicate rows:")
print(df_users.duplicated().sum())

🔹 Missing values:
user_id             0
birthdate           0
gender              0
married             0
has_children        0
home_country        0
home_city           0
home_airport        0
home_airport_lat    0
home_airport_lon    0
sign_up_date        0
dtype: int64

🔹 Number of duplicate rows:
0


# Transformation

## - flights